docs/dev/framework/routing/content-routing.md

```diff
@@ -0,0 +1,301 @@
+---
+title: Content Routing
+description: "URL generation for database records in Contao."
+---
+
+
+{{< version "5.3" >}}
+
+In Symfony you can generate the URL to the routes of controllers via the [`router` service][RouterService]. However,
+in Contao you also want to generate front end URLs for your database records (like a news item for example) - or just
+for a regular page.
+
+Contao **5.3** introduces a system with which you can implement "URL resolvers" for your own database records (models,
+entities) and then use the `ContentUrlGenerator` service to generate URLs to these database records - or to any of the
+other existing models in Contao, like news or pages.
+
+
+## Content URL Generation
+
+Since Contao **5.3** this can be done via the `ContentUrlGenerator` service, which works analogous to the Symfony
+router service. It has a `generate()` method, just like the Symfony URL generator - but instead of a route name it
+expects an object for which the URL should be generated as its first parameter. You can also optionally pass parameters
+for the URL generation and also define the URL reference type (e.g. absolute URL, absolute path, etc.).
+
+This is important for when your front end module references a page as the redirect target for a form for example. Or
+if your front end module creates a news list and you need the URL to the detail page of each news item.
+
+```php
+// src/MyService.php
+use Contao\CoreBundle\Routing\ContentUrlGenerator;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+
+class MyService
+{
+    public function __construct(private readonly ContentUrlGenerator $contentUrlGenerator)
+    {
+    }
+
+    public function __invoke()
+    {
+        $page = PageModel::findBy(…);
+
+        // Generates an absolute URL for the given page
+        $pageUrl = $this->contentUrlGenerator->generate($page, [], UrlGeneratorInterface::ABSOLUTE_URL);
+
+        $news = NewsModel::findBy(…);
```

Comment:
This should use the framework, I don't think advocating for static calls here is a good idea. Maybe even better would be to have a fragment example which already gets the model in the `__invoke()`.

Reply:
> This should use the framework, I don't think advocating for static calls here is a good idea.

If you don't need tests I see no reason to use the adapters of the framework. Besides we use `Model::findBy` numerous times in the documentation already. I think it would be superfluous to change this to adapters everywhere.

> Maybe even better would be to have a fragment example which already gets the model in the __invoke().

Hm, what would such an example look like though? Typically you will need to generate the detail URL of an object in a module that lists several such object.

And I am not even sure it is technically possible to have parameters automatically mapped like that anyway?

Reply:
> Hm, what would such an example look like though?

A list controller? That's the whole point of this example?

Reply:
How would you get the models for a list controller in the `__invoke()` method?

Reply:
Regardless, I will add an example with a list controller where you generate the URL in the twig template - that is definitely still missing.

Reply:
I guess based on the settings of the fragment. I mean, in the news list controller I can configure the settings so that's what you have to pass on to your `NewsModel::findBy($criteria)`.

Reply:
Ah yes ok. Well I have included a simple example now.

Reply:
> If you don't need tests I see no reason to use the adapters of the framework

Fully agree with this (and stick to this in all of my extensions as well).